stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,523 @@
+     * @param dest points to values holding adjoint accumulate.
+     * @param x An eigen type holding vars to accumulate over.
+     * @param args Further args to accumulate over.
+     */
+    template <typename EigT, require_eigen_vt<is_var, EigT>* = nullptr,
+              typename... Pargs>
+    inline double* accumulate_adjoints(double* dest, EigT&& x,
+                                       Pargs&&... args) {
+      Eigen::Map<Eigen::MatrixXd>(dest, x.rows(), x.cols()) += x.adj();
+      return accumulate_adjoints(dest + x.size(), std::forward<Pargs>(args)...);
+    }
+
+    /**
+     * Specialization that is a no-op for Arithmetic types.
+     * @tparam Arith A type satisfying `std::is_arithmetic`.
+     * @tparam Pargs types of further arguments to accumulate over.
+     * @param dest points to values holding adjoint accumulate.
+     * @param x An object that is either arithmetic or a container of Arithmetic
+     *  types.
+     * @param args Further args to accumulate over.
+     */
+    template <typename Arith,
+              require_arithmetic_t<scalar_type_t<Arith>>* = nullptr,
+              typename... Pargs>
+    inline double* accumulate_adjoints(double* dest, Arith&& x,
+                                       Pargs&&... args) {
+      return accumulate_adjoints(dest, std::forward<Pargs>(args)...);
+    }
+
+    inline double* accumulate_adjoints(double* x) { return x; }
+
+    /**
+     * Within thread reduction.
+     */
+    inline void operator()(const tbb::blocked_range<size_t>& r) {
+      if (r.empty()) {
+        return;
+      }
+
+      if (args_adjoints_.size() == 0) {
+        args_adjoints_ = Eigen::VectorXd::Zero(this->num_shared_terms_);
+      }
+
+      const nested_rev_autodiff begin_nest;
+      // create a deep copy of all var's so that these are not
+      // linked to any outer AD tree
+      std::decay_t<Vec> local_sub_slice;
+      local_sub_slice.reserve(r.size());
+      for (int i = r.begin(); i < r.end(); ++i) {
+        local_sub_slice.emplace_back(deep_copy(vmapped_[i]));
+      }
+      auto args_tuple_local_copy = apply(
+          [&](auto&&... args) {
+            return std::tuple<decltype(deep_copy(args))...>(deep_copy(args)...);
+          },
+          this->args_tuple_);
+      var sub_sum_v = apply(
+          [&](auto&&... args) {
+            return ReduceFunction()(r.begin(), r.end() - 1, local_sub_slice,
+                                    this->msgs_, args...);
+          },
+          args_tuple_local_copy);
+      sub_sum_v.grad();
+      sum_ += sub_sum_v.val();
+      accumulate_adjoints(
+          this->sliced_partials_ + r.begin() * per_job_sliced_terms_,
+          local_sub_slice);
+      apply(
+          [&](auto&&... args) {
+            return accumulate_adjoints(args_adjoints_.data(),
+                                       std::forward<decltype(args)>(args)...);
+          },
+          std::move(args_tuple_local_copy));
```

Comment:
Out of curiosity: What's the intent of the move here? Shouldn't a forward have the same effect?

Reply:
Also why do we need anything? Why wouldn't the auto&& just absorb the right thing?

Reply:
The intent of the move here is to tell the compiler "You can move ownership of the memory from `args_tuple_local_copy` to this function". Is it worth it? Eh, I mean if it's too confusing we can get rid of it. It mostly just enables use of move semantics

> Also why do we need anything? 

tbh a fair question. With this stuff I was trying to move and utilize temporaries as best as possible since our biggest bottle neck here is memory overhead

> Why wouldn't the auto&& just absorb the right thing?

Unless you tell the compiler that it's okay to transfer ownership of memory it will not. The godbolt example here shows that. 
https://godbolt.org/z/Q8aFnc

 For more on this you can check out C++ Templates, The Complete Guide Chapter 7 of the second edition. Effective Modern C++ also has a chapter on this stuff though I can't remember which.

Reply:
> Unless you tell the compiler that it's okay to transfer ownership of memory it will not

Ooooooo. That example is a bit confusing, so many puts

Reply:
Oh so after you move it in there, it wouldn't be safe to mess with it anymore? That function will destruct it and everything?

Reply:
Oh really? For future reference can you think of an easier way to show that? I mostly wanted the puts so the output on the right side looked pretty

```
------
S b(a) uses copy constructor
Copy Constructor
------
S c(std::move(a)) uses move constructor
Move Constructor
------
```

> Oh so after you move it in there, it wouldn't be safe to mess with it anymore? That function will destruct it and everything?

Yes doing a `move` leaves the original object in a valid but undefined state. You only want to do it at the end of a function when you know you won't need that object anymore. There's a nice blog post [here](https://medium.com/@winwardo/c-moves-for-people-who-dont-know-or-care-what-rvalues-are-%EF%B8%8F-56ee122dda7) that covers a lot of what move actually does.

Reply:
OOoh, I see, I was trying to decode the assembly to see where there was memory being shared which isn't the point.

It's showing me how we get the special move constructor that can be coded to do different things (like take ownership of memory). My bad.

> There's a nice blog post

Thanks! I'll read it. But be warned I'll probably keep forgetting and asking anyway lol.

Reply:
> OOoh, I see, I was trying to decode the assembly to see where there was memory being shared which isn't the point.

Lol oh sorry I should have just deleted that